README.md
# Twilio Voice plugin for Koha

This plugin allows Koha to send 'phone' notices via the Twilio service API.

# Installation

This plugin requires the installation of following Perl modules on the Koha server:
* `WWW::Twilio::TwiML`
* `WWW::Form::UrlEncoded`

If those modules are not installed, this plugin may not appear in your Koha instance's list of installed plugins.

# Koha configuration
* Ensure the syspref `PhoneNotification` is enabled.
* Ensure the syspref `TalkingTechItivaPhoneNotification` is *not* enabled.
* Create the `phone` transport version of all notices patron's can opt to recieve via phone.
  * These notices will be read as-is over the phone via Twilio.
  * Best to keep notices as terse as possible.
  * Repeat the entire notice twice in case the call goes to voicemail.

# Plugin Configuration

> :warning: **Do not call your patrons by accident in the middle of the night!** Run multiple instances of `process_message_queue.pl` with the `-t` paramter. Limit the sending of phone calls to common waking hours in your area.

* Create a Twilio account, log in to Twilio
* Create a new project
* Get a trial number, or verify an existing phone number for this project
* Verify your phone number for testing purposes 
* Note your Account SID and Auth Token on the project landing page
* Browse to the Twilio plugin configuration page in Koha
* From there, you can plug in the Account SID, Auth Token, and Twilio phone number from  your Twilio project


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.6] - 2021-06-29
### Added
- Add ability to detect "type" command line parameter and honor it
- Add ability to detect "letter code" command line parameter and honor it

## [0.0.9] - 2021-02-01
### Added
- Add ability to send only first of HOLD notices for each patron

## [0.0.5] - 2021-01-15
### Changed
- Don't reset any call status back to "pending", it may be causing duplicate calls.

## [0.0.2] - 2020-05-13
### Added
- This plugin, initial commit ;)


Koha/Plugin/Com/ByWaterSolutions/TwilioVoice/API.pm
package Koha::Plugin::Com::ByWaterSolutions::TwilioVoice::API;

# This file is part of Koha.
#
# Koha is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Koha; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

use Modern::Perl;

use Mojo::Base 'Mojolicious::Controller';

use WWW::Twilio::TwiML;
use WWW::Form::UrlEncoded qw(parse_urlencoded);

use Koha::Notice::Messages;

=head1 API

=head2 Class Methods

=head3 Returns TwiML for the given message

=cut

sub twiml {
    my $c = shift->openapi->valid_input or return;

    my $message_id = $c->validation->param('message_id');
    my $message    = Koha::Notice::Messages->find($message_id);
    unless ($message) {
        return $c->render(
            status  => 404,
            openapi => { error => "Message not found." }
        );
    }

    my $tw = new WWW::Twilio::TwiML;
    $tw->Response
    ->Pause({length => 2})->parent
    ->Say(
        {
            voice    => "Polly.Joanna",
            language => "en-US"
        },
        $message->content
    )->parent
    ->Pause({length => 2})->parent
    ->Say(
        {
            voice    => "Polly.Joanna",
            language => "en-US"
        },
        $message->content
    );

    $message->status('sent');
    $message->store();

    warn "TWILIO: twiml(): " . Data::Dumper::Dumper( $tw->to_string );

    return $c->render( status => 200, format => "xml", text => $tw->to_string );
}

sub update_message_status {
    my $c = shift->openapi->valid_input or return;

    my $message_id = $c->validation->param('message_id');
    my $message    = Koha::Notice::Messages->find($message_id);
    unless ($message) {
        return $c->render(
            status  => 404,
            openapi => { error => "Message not found." }
        );
    }

    my $body = $c->req->body;

    if ( my %data = parse_urlencoded($body) ) {
        my $twilio_status = $data{CallStatus};

        warn "TWILIO: update_message_status(): " . Data::Dumper::Dumper( \%data );

        my $status = $twilio_status eq 'queued'      ? 'sent'   : # We should get another status update later
                     $twilio_status eq 'ringing'     ? 'sent'   : # Ditto
                     $twilio_status eq 'in-progress' ? 'sent'   : # The person picked up, basically completed
                     $twilio_status eq 'completed'   ? 'sent'   : # Clearly completed
                     $twilio_status eq 'busy'        ? 'failed' : # TODO: Make retrying busy a plugin setting
                     $twilio_status eq 'failed'      ? 'failed' : # Phone number was most likely invalid
                     $twilio_status eq 'no-answer'   ? 'failed' : # See TODO above
                                                       'failed' ; # Staus was something we didn't expect
        $message->status($status);
        $message->store();

        return $c->render( status => 204, text => q{} );
    }
    else {
        return $c->render(
            status  => 500,
            openapi => { error => "Unable to decode json" }
        );
    }
}

1;


Koha/Plugin/Com/ByWaterSolutions/TwilioVoice/BotherPatronController.pm
package Koha::Plugin::Com::ByWaterSolutions::TwilioVoice::BotherPatronController;

# This file is part of Koha.
#
# Koha is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Koha; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

use Modern::Perl;

use Mojo::Base 'Mojolicious::Controller';

=head1 API

=head2 Class Methods

=head3 Method that bothers the patron, with no side effects

=cut

sub bother {
    my $c = shift->openapi->valid_input or return;

    my $patron_id = $c->validation->param('patron_id');
    my $patron    = Koha::Patrons->find($patron_id);

    unless ($patron) {
        return $c->render( status => 404, openapi => { error => "Patron not found." } );
    }

    return $c->render( status => 200, openapi => { bothered => Mojo::JSON->true } );
}

1;


Koha/Plugin/Com/ByWaterSolutions/TwilioVoice.pm
package Koha::Plugin::Com::ByWaterSolutions::TwilioVoice;

use Modern::Perl;

use base qw(Koha::Plugins::Base);

use C4::Auth;
use C4::Context;
use Koha::Notice::Messages;

use HTTP::Request::Common;
use LWP::UserAgent;
use Mojo::JSON qw(decode_json);

## Here we set our plugin version
our $VERSION         = "{VERSION}";
our $MINIMUM_VERSION = "19.11.06";

## Here is our metadata, some keys are required, some are optional
our $metadata = {
    name            => 'Twilio Voice Plugin',
    author          => 'Kyle M Hall',
    date_authored   => '2020-05-13',
    date_updated    => "1900-01-01",
    minimum_version => $MINIMUM_VERSION,
    maximum_version => undef,
    version         => $VERSION,
    description =>
      'This plugin enables sending of phone message to patrons via Twilio.',
};

sub new {
    my ( $class, $args ) = @_;

    ## We need to add our metadata here so our base class can access it
    $args->{'metadata'} = $metadata;
    $args->{'metadata'}->{'class'} = $class;

    ## Here, we call the 'new' method for our base class
    ## This runs some additional magic and checking
    ## and returns our actual $self
    my $self = $class->SUPER::new($args);

    return $self;
}

sub before_send_messages {
    my ($self, $params) = @_;

    my $type        = $params->{type};
    my $letter_code = $params->{letter_code};

    # If a type limit is passed in, only run if the type is "phone"
    return if $type && $type ne 'phone';

    my $AccountSid = $self->retrieve_data('AccountSid');
    my $AuthToken  = $self->retrieve_data('AuthToken');
    my $single_notice_hold = $self->retrieve_data('single_notice_hold');
    my $skip_if_other_transports = $self->retrieve_data('skip_if_other_transports');

    my $from = $self->retrieve_data('From');

    my $parameters = {
        status                 => 'pending',
        message_transport_type => 'phone',
    };
    $params->{letter_code} = $letter_code if $letter_code;
    my $messages = Koha::Notice::Messages->search($parameters);

    my $sent = {};
    while ( my $m = $messages->next ) {
        $m->status('sent');
        $m->update();

        if ( $m->letter_code eq 'HOLD' && $single_notice_hold ) {
            if ( $sent->{HOLD}->{ $m->borrowernumber } ) {
                $sent->{HOLD}->{ $m->borrowernumber } = 1;

                $m->status('deleted'); # As close a status to 'skipped' as we have
                $m->update();

                next;
            }
        }

        if ( $skip_if_other_transports ) {
            my $other_messages = Koha::Notice::Messages->search(
                {
                    -and => [
                        borrowernumber => $m->borrowernumber,
                        status => 'pending',
                        letter_code => $m->letter_code,
                        -or => [
                            message_transport_type => 'email',
                            message_transport_type => 'sms',
                        ]
                    ],
                }
            );

            if ( $other_messages->count ) {
                $m->status('deleted'); # As close a status to 'skipped' as we have
                $m->update();

                next;
            }
        }

        my $patron = Koha::Patrons->find( $m->borrowernumber );
        next unless $patron;

        my $phone = $patron->phone || $patron->mobile;

        # Normalize the phone number to E.164 format, Twilio has a convenient ( and free ) API for this.
        my $ua = LWP::UserAgent->new;
        my $request =
          HTTP::Request->new( GET =>
              "https://lookups.twilio.com/v1/PhoneNumbers/$phone?CountryCode=US"
          );
        $request->authorization_basic( $AccountSid, $AuthToken );
        my $response = $ua->request($request);
        next if $response->code eq "404";
        my $data = decode_json( $response->decoded_content );
        my $to   = $data->{phone_number};

        my $staffClientBaseURL = $self->retrieve_data('IncomingApiCallsUrl') || C4::Context->preference('staffClientBaseURL');
        $staffClientBaseURL =~ s/[^[:print:]]+//g;
        $staffClientBaseURL =~ s/[^[:ascii:]]+//g;

        # Send the call request
        my $message_id = $m->id;
        my $url = "https://api.twilio.com/2010-04-01/Accounts/$AccountSid/Calls.json";
        my $twiml_url = "$staffClientBaseURL/api/v1/contrib/twiliovoice/message/$message_id/twiml";
        my $status_callback_url = "$staffClientBaseURL/api/v1/contrib/twiliovoice/message/$message_id/status";

        warn "Twilio Phone message sent to $to for message id $message_id";

        $request = POST $url,
          [
            From                 => $from,
            To                   => $to,
            Url                  => $twiml_url,
            StatusCallback       => $status_callback_url,
            StatusCallbackEvent  => 'completed',
            StatusCallbackMethod => 'POST',
            MachineDetection     => 'DetectMessageEnd',
          ];
        $request->authorization_basic( $AccountSid, $AuthToken );
        $response = $ua->request($request);

        unless ($response->is_success) {
            warn "Twilio response indicates failure: " . $response->status_line;
        }
    }
}

sub configure {
    my ( $self, $args ) = @_;
    my $cgi = $self->{'cgi'};

    unless ( $cgi->param('save') ) {
        my $template = $self->get_template( { file => 'configure.tt' } );

        ## Grab the values we already have for our settings, if any exist
        $template->param(
            AccountSid => $self->retrieve_data('AccountSid'),
            AuthToken  => $self->retrieve_data('AuthToken'),
            From       => $self->retrieve_data('From'),
            IncomingApiCallsUrl  => $self->retrieve_data('IncomingApiCallsUrl'),
            single_notice_hold => $self->retrieve_data('single_notice_hold'),
            skip_if_other_transports => $self->retrieve_data('skip_if_other_transports'),
        );

        $self->output_html( $template->output() );
    }
    else {
        $self->store_data(
            {
                AccountSid => $cgi->param('AccountSid'),
                AuthToken  => $cgi->param('AuthToken'),
                From       => $cgi->param('From'),
                IncomingApiCallsUrl  => $cgi->param('IncomingApiCallsUrl'),
                single_notice_hold => $cgi->param('single_notice_hold') ? 1 : 0,
                skip_if_other_transports => $cgi->param('skip_if_other_transports') ? 1 : 0,
            }
        );
        $self->go_home();
    }
}

sub install() {
    my ( $self, $args ) = @_;

    return 1;
}

sub upgrade {
    my ( $self, $args ) = @_;

    return 1;
}

sub uninstall() {
    my ( $self, $args ) = @_;

    return 1;
}

sub api_routes {
    my ( $self, $args ) = @_;

    my $spec_str = $self->mbf_read('openapi.json');
    my $spec     = decode_json($spec_str);

    return $spec;
}

sub api_namespace {
    my ($self) = @_;

    return 'twiliovoice';
}

1;
